### PR TITLE
feat: origin detection with container ID field

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

Adds `c:<id>` field holding the container ID so the agent can attach container tags.

### Description of the Change

- If `DD_ENTITY_ID` is set, we ignore the container field
- If `DD_ENTITY_ID` is not set and `DD_ORIGIN_DETECTION_ENABLED` is not explicitly set to false and the `origin_detection_enabled` param is not explicitly disabled, we try to get the app container ID by parsing `/proc/self/cgroup`. In case of success, the container field in will be set.

### Verification Process

Wrote a small app importing the lib and sending metrics to agent 7.35 with origin detection client enabled - the metrics were tagged with container tags.

### Additional Notes

- Users should explicitly enable container tagging in the agent otherwise the agent ignores the container field. So even if the client sends the field automatically, the tagging won't be done without explicitly enabling it in the agent.
- Cgroup parsing logic is essentially the same as in [dd-trace-py](https://github.com/DataDog/dd-trace-py/pull/1007/files)
- This feature requires an agent change https://github.com/DataDog/datadog-agent/pull/10659
- Reference implementation is in Go https://github.com/DataDog/datadog-go/pull/250
- Public docs https://github.com/DataDog/documentation/pull/13241

### Release Notes

Support adding the container ID field for origin detection (dogstatsd 1.2)

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/datadogpy/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: datadog/dogstatsd/base.py [hunk 1]
```diff
@@ -274,10 +308,13 @@ def __init__(
         # Options
         env_tags = [tag for tag in os.environ.get("DATADOG_TAGS", "").split(",") if tag]
         # Inject values of DD_* environment variables as global tags.
+        has_entity_id = False
         for var, tag_name in DD_ENV_TAGS_MAPPING.items():
             value = os.environ.get(var, "")
             if value:
                 env_tags.append("{name}:{value}".format(name=tag_name, value=value))
+                if value == ENTITY_ID_ENV_VAR:
```
Comment (vickenty) [inline datadog/dogstatsd/base.py:316]: Did you mean to write `var` here?

Comment (ahmed-mez) [reply datadog/dogstatsd/base.py:316]: Nice catch! thanks

File: datadog/dogstatsd/container.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                        if match:
+                            return match.group(1)
+        except IOError as e:
+            if e.errno != errno.ENOENT:
+                raise NotImplementedError("Unable to open {}.".format(self.CGROUP_PATH))
+        except Exception:
+            raise UnresolvableContainerID("Unable to read the container ID.")
+        return None
+
+    def get_container_id(self):
```
Comment (sgnn7) [inline datadog/dogstatsd/container.py:59]: Mentioned before but it's probably better to have this as a property getter

Comment (sgnn7) [inline datadog/dogstatsd/container.py:59]: Mentioned before but it's probably better to have this as a property getter

**Review by sgnn7 — COMMENTED** [review]
@ahmed-mez Left some comments but overall looks good

File: datadog/dogstatsd/base.py [hunk 3]
```diff
@@ -52,7 +59,7 @@
 
 # Mapping of each "DD_" prefixed environment variable to a specific tag name
 DD_ENV_TAGS_MAPPING = {
-    "DD_ENTITY_ID": ENTITY_ID_TAG_NAME,
+    ENTITY_ID_ENV_VAR: ENTITY_ID_TAG_NAME,
```
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply datadog/dogstatsd/base.py:62]: I don't think so, the value of the `ENTITY_ID_ENV_VAR` var is `"DD_ENTITY_ID"`. We need the value of the variable not the string `"ENTITY_ID_ENV_VAR"`.

File: datadog/dogstatsd/base.py [hunk 4]
```diff
@@ -206,6 +219,27 @@ def __init__(
         :param telemetry_socket_path: Submit client telemetry to dogstatsd through a UNIX
         socket instead of UDP. If set, disables UDP transmission (Linux only)
         :type telemetry_socket_path: string
+
+        :param container_id: Allows passing the container ID, this will be used by the Agent to enrich
+        metrics with container tags.
+        This feature requires Datadog Agent version >=6.35.0 && <7.0.0 or Agent versions >=7.35.0.
+        When configured, the provided container ID is prioritized over the container ID discovered
+        via Origin Detection.
+        The client prioritizes the value passed via DD_ENTITY_ID (if set) over the container ID.
```
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply datadog/dogstatsd/base.py:228]: Not really. `the provided container ID` and `the value passed via DD_ENTITY_ID` are two different things, the former is the container ID provided as argument, the latter is the pod uid injected as env var via the kubernetes downward api.

[the remark this answers is not in the document]
Comment (ahmed-mez) [reply datadog/dogstatsd/base.py:228]: Not really. `the provided container ID` and `the value passed via DD_ENTITY_ID` are two different things, the former is the container ID provided as argument, the latter is the pod uid injected as env var via the kubernetes downward api.

File: datadog/dogstatsd/base.py [hunk 5]
```diff
@@ -717,13 +762,14 @@ def close_socket(self):
...
         # Create/format the metric packet
-        return "%s%s:%s|%s%s%s" % (
+        return "%s%s:%s|%s%s%s%s" % (
             (self.namespace + ".") if self.namespace else "",
             metric,
             value,
             metric_type,
             ("|@" + text(sample_rate)) if sample_rate != 1 else "",
             ("|#" + ",".join(normalize_tags(tags))) if tags else "",
+            self._get_container_id_field(),
```
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply datadog/dogstatsd/base.py:772]: No strong opinion, but I made this decision so we can do the (`|c:` + `container_id`) operation only once and avoid duplicating the `if container_id: append..` for metrics, events and service checks.
I'll use a property get instead 👍

File: datadog/dogstatsd/base.py [hunk 6]
```diff
@@ -970,6 +1017,7 @@ def service_check(
             string = u"{0}|#{1}".format(string, ",".join(tags))
         if message:
             string = u"{0}|m:{1}".format(string, message)
+        string = u"{0}{1}".format(string, self._get_container_id_field())
```
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply datadog/dogstatsd/base.py:1020]: I explained why I used a method in [this comment](https://github.com/DataDog/datadogpy/pull/720#discussion_r821469798). Let's choose consistency 😅

File: datadog/dogstatsd/base.py [hunk 7]
```diff
@@ -984,5 +1032,48 @@ def _add_constant_tags(self, tags):
...
+        If DD_ENTITY_ID is not set, we try to fill the container field automatically unless
+        DD_ORIGIN_DETECTION_ENABLED is explicitly set to false.
+        """
+        if not origin_detection_enabled or has_entity_id or container_id is not None:
+            # origin detection is explicitly disabled
+            # or DD_ENTITY_ID was found
+            # or a user-defined container ID was provided
+            return False
+        value = os.environ.get(ORIGIN_DETECTION_ENABLED, "")
+        return value.lower() not in {"no", "false", "0", "n", "off"}
```
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply datadog/dogstatsd/base.py:1050]: Would you mind elaborating?
I know that set size here is very small and the operation is not in a hot path, but I'd prefer to use a set for the O(1) lookup. Am I missing something?

[the remark this answers is not in the document]
Comment (ahmed-mez) [reply datadog/dogstatsd/base.py:1052]: Would you mind elaborating?
I know that set size here is very small and the operation is not in a hot path, but I'd prefer to use a set for the O(1) lookup. Am I missing something?

File: datadog/dogstatsd/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sgnn7) [reply datadog/dogstatsd/base.py:62]: Good point - I don't know why I missed the fact that this was the varname from above.

**Review by sgnn7 — APPROVED** [review]

**Pull request merged**